### ISSUE TITLE
planner/cascades: extract `case when` from AggFunc as a Selection

### ISSUE THREAD

**Issue by francis0407** [pingcap/tidb#14042]
## Feature Request

**Describe the feature you'd like:**

Aggregate queries contain `case when` expression sometimes can be converted to a Aggregation with a Selection whicn can be pushed down.
For example, we have a SQL: `select sum(case when a > 10 then b) from t`
Currently, TiDB may generate such a plan:
```
mysql> explain select sum(case when a > 3 then b end) from t;
+------------------------+-------+-----------+---------------------------------------------------------------------+
| id                     | count | task      | operator info                                                       |
+------------------------+-------+-----------+---------------------------------------------------------------------+
| StreamAgg_24           | 1.00  | root      | funcs:sum(Column#13)->Column#5                                      |
| └─IndexReader_25       | 1.00  | root      | index:StreamAgg_8                                                   |
|   └─StreamAgg_8        | 1.00  | cop[tikv] | funcs:sum(case(gt(test.t.a, 3), test.t.b))->Column#13               |
|     └─IndexScan_23     | 5.00  | cop[tikv] | table:t, index:b, range:[NULL,+inf], keep order:false, stats:pseudo |
+------------------------+-------+-----------+---------------------------------------------------------------------+
4 rows in set (0.00 sec)
```

But actually, it is equivalent to `select sum(b) from t where a > 3` which can uses the index of 'a'.
```
mysql> explain select sum(b) from t where a > 3;
+------------------------+-------+-----------+---------------------------------------------------------+
| id                     | count | task      | operator info                                           |
+------------------------+-------+-----------+---------------------------------------------------------+
| StreamAgg_17           | 1.00  | root      | funcs:sum(Column#7)->Column#5                           |
| └─TableReader_18       | 1.00  | root      | data:StreamAgg_9                                        |
|   └─StreamAgg_9        | 1.00  | cop[tikv] | funcs:sum(test.t.b)->Column#7                           |
|     └─TableScan_16     | 2.08  | cop[tikv] | table:t, range:(3,+inf], keep order:false, stats:pseudo |
+------------------------+-------+-----------+---------------------------------------------------------+
4 rows in set (0.01 sec)
```
**Describe alternatives you've considered:**

Apache Calcite implements a [rule](https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/rel/rules/AggregateCaseToFilterRule.java) which could be a good reference. 

And we'd better implement this in [cascades planner](https://github.com/pingcap/tidb/projects/16).

**Comment by SeaRise** [comment]
I can try it

**Comment by francis0407** [comment]
Closed #14526
